server/block/campfire.go

```diff
@@ -0,0 +1,256 @@
+	return newBreakInfo(5, alwaysHarvestable, axeEffective, func(t item.Tool, enchantments []item.Enchantment) []item.Stack {
+		var drops []item.Stack
+		if hasSilkTouch(enchantments) {
+			drops = append(drops, item.NewStack(c, 1))
+		} else {
+			switch c.Type {
+			case NormalFire():
+				drops = append(drops, item.NewStack(item.Charcoal{}, 2))
+			case SoulFire():
+				drops = append(drops, item.NewStack(SoulSoil{}, 1))
+			}
+		}
+		for _, v := range c.Items {
+			if !v.Item.Empty() {
+				drops = append(drops, v.Item)
+			}
+		}
+		return drops
+	})
+}
+
+// LightEmissionLevel ...
+func (c Campfire) LightEmissionLevel() uint8 {
+	if c.Extinguished {
+		return 0
+	}
+	return c.Type.LightLevel()
+}
+
+// Ignite ...
+func (c Campfire) Ignite(pos cube.Pos, w *world.World) bool {
+	w.PlaySound(pos.Vec3(), sound.Ignite{})
+	if _, ok := w.Liquid(pos); ok {
+		return true
+	}
+
+	c.Extinguished = false
+	w.SetBlock(pos, c, nil)
+	return true
+}
+
+// Splash ...
+func (c Campfire) Splash(w *world.World, pos cube.Pos, p potion.Potion) {
+	if p != potion.Water() {
+		// Water is the only potion that can extinguish a campfire.
+		return
+	}
+
+	w.PlaySound(pos.Vec3Centre(), sound.FireExtinguish{})
+	c.Extinguished = true
+	w.SetBlock(pos, c, nil)
+}
+
+// Activate ...
+func (c Campfire) Activate(pos cube.Pos, _ cube.Face, w *world.World, u item.User, ctx *item.UseContext) bool {
+	if held, _ := u.HeldItems(); !held.Empty() {
+		if rawFood, ok := held.Item().(item.Smeltable); ok && rawFood.SmeltInfo().Food {
+			for i, it := range c.Items {
+				if it.Item.Empty() {
+					c.Items[i] = CampfireItem{
+						Item: held.Grow(-held.Count() + 1),
+						Time: time.Second * 30,
+					}
+
+					ctx.SubtractFromCount(1)
+
+					w.PlaySound(pos.Vec3Centre(), sound.ItemAdd{})
+					w.SetBlock(pos, c, nil)
+					return true
+				}
+			}
+		}
+	}
+	return false
+}
+
+// UseOnBlock ...
+func (c Campfire) UseOnBlock(pos cube.Pos, face cube.Face, _ mgl64.Vec3, w *world.World, user item.User, ctx *item.UseContext) (used bool) {
+	if _, ok := w.Block(pos).(Campfire); ok && face == cube.FaceUp {
+		return false
+	}
+	pos, _, used = firstReplaceable(w, pos, face, c)
+	if !used {
+		return
+	}
+	c.Facing = user.Facing().Opposite()
+	place(w, pos, c, user, ctx)
+	return placed(ctx)
+}
+
+// Tick is called to cook the items within the campfire.
+func (c Campfire) Tick(_ int64, pos cube.Pos, w *world.World) {
+	if c.Extinguished {
+		// Extinguished, do nothing.
+		return
+	}
+	if rand.Float64() <= 0.016 { // Every three or so seconds.
+		w.PlaySound(pos.Vec3Centre(), sound.CampfireCrackle{})
+	}
+
+	for i, it := range c.Items {
+		if !it.Item.Empty() && it.Time <= 0 {
+			if food, ok := it.Item.Item().(item.Smeltable); ok {
+				dropItem(w, food.SmeltInfo().Product, pos.Vec3Middle())
+			}
+
+			c.Items[i].Item = item.Stack{}
+			w.SetBlock(pos, c, nil)
+			continue
+		}
+		c.Items[i].Time = it.Time - time.Millisecond*50
+		w.SetBlock(pos, c, nil)
```

Comment:
```go
c.Items[i].Time = it.Time - time.Millisecond*50
w.SetBlock(pos, c, nil)
```
Because c.Items is an array, this code is called 4 times every tick for campfires when they have nothing on them. Consider adding a `if it.Empty() { continue }` at the start.